File: javascript/src/billing/postgres/repositories/customer.ts
import { z } from "zod";
import type { QueryFn } from "../../../shared/postgres-types.js";
import {
  optionalRecordRow,
  postgresUuid,
  requireResultField,
  safeParse,
} from "../../../shared/postgres-validation.js";

const CustomerRowSchema = z.object({
  subject_id: postgresUuid,
});

const CustomerByUserRowSchema = z.object({
  provider: z.string().min(1),
  provider_customer_id: z.string().min(1),
});

export class BillingCustomerRepository {
  constructor(private query: QueryFn) {}

  async upsert(
    provider: string,
    providerCustomerId: string,
    userId: string,
    email: string | null,
  ): Promise<void> {
    const rows = await this.query(
      "SELECT bursar.upsert_billing_customer($1::uuid, $2, $3, $4) AS id",
      [userId, provider, providerCustomerId, email],
    );
    requireResultField(rows, "id", postgresUuid, "BillingCustomerRepository.upsert");
  }

  async get(provider: string, providerCustomerId: string): Promise<string | null> {
    const rows = await this.query("SELECT * FROM bursar.get_billing_customer_by_provider($1, $2)", [
      provider,
      providerCustomerId,
    ]);
    const row = optionalRecordRow(rows, "BillingCustomerRepository.get");
    return row === null
      ? null
      : safeParse(CustomerRowSchema, row, "BillingCustomerRepository.get").subject_id;
  }

  async getByUserId(
    userId: string,
    provider?: string | null,
  ): Promise<{ provider: string; providerCustomerId: string } | null> {
    const rows = await this.query("SELECT * FROM bursar.get_billing_customer($1::uuid, $2)", [
      userId,
      provider ?? null,
    ]);
    const row = optionalRecordRow(rows, "BillingCustomerRepository.getByUserId");
    if (row === null) return null;
    const parsed = safeParse(CustomerByUserRowSchema, row, "BillingCustomerRepository.getByUserId");
    return { provider: parsed.provider, providerCustomerId: parsed.provider_customer_id };
  }
}

File: python/src/bursar/billing/postgres/repositories/customer.py
from __future__ import annotations

from uuid import UUID

from bursar.billing.types import BillingCustomerRecord
from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories._utils import (
    optional_mapping_row,
    require_identifier_result,
    validate_non_empty,
)
from bursar.errors import StoreError


class BillingCustomerRepository:
    def __init__(self, execute: DbQuery) -> None:
        self._execute = execute

    def upsert(self, provider: str, provider_customer_id: str, user_id: str, email: str | None) -> None:
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_customer_id, "provider_customer_id")
        rows = self._execute(
            "SELECT bursar.upsert_billing_customer(%s::uuid, %s, %s, %s) AS id",
            [user_id, provider, provider_customer_id, email],
        )
        require_identifier_result(rows, "id", "BillingCustomerRepository.upsert")

    def get(self, provider: str, provider_customer_id: str) -> str | None:
        rows = self._execute(
            "SELECT * FROM bursar.get_billing_customer_by_provider(%s, %s)",
            [provider, provider_customer_id],
        )
        row = optional_mapping_row(rows, "BillingCustomerRepository.get")
        if row is None:
            return None
        try:
            subject_id = str(UUID(str(row.get("subject_id"))))
        except (AttributeError, TypeError, ValueError) as error:
            raise StoreError("BillingCustomerRepository.get: malformed customer row") from error
        return subject_id

    def get_by_user_id(self, user_id: str, provider: str | None = None) -> BillingCustomerRecord | None:
        rows = self._execute("SELECT * FROM bursar.get_billing_customer(%s::uuid, %s)", [user_id, provider])
        row = optional_mapping_row(rows, "BillingCustomerRepository.get_by_user_id")
        if row is None:
            return None
        if (
            not isinstance(row, dict)
            or not isinstance(row.get("provider"), str)
            or not row["provider"]
            or not isinstance(row.get("provider_customer_id"), str)
            or not row["provider_customer_id"]
        ):
            raise StoreError("BillingCustomerRepository.get_by_user_id: malformed customer row")
        return BillingCustomerRecord(
            provider=row["provider"],
            provider_customer_id=row["provider_customer_id"],
        )
